rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp

```diff
@@ -143,9 +143,10 @@ bool SequentialReader::has_next()
     if (!current_storage_has_next && has_next_file()) {
       load_next_file();
       storage_options_.uri = get_current_file();
-
+      // open and set filters
       storage_ = storage_factory_->open_read_only(storage_options_);
-      storage_->set_filter(topics_filter_);
+      storage_->seek(seek_time_);
```

Comment:
@MichaelOrlov for the same reason that we had `->set_filter(topics_filter_)`. `seek_time_` is basically just a time filter in the query, and when you roll over to a new file, you need to setup the filters of the query to be run in the new storage reader.

The `sequential_reader` kind of assumes that the files in a split bag are arranged in non-overlapping time order, so setting the `seek_time_` is redundant given that assumption, but there isn't actually anything that reinforces this assumption, I can easily create a multi-file bag that breaks it, and the reader wouldn't know any better (see #842). While setting the `seek_time_` here wouldn't necessarily resolve this assumption break, it will ensure that reads across multiple files will consistently obey the seek time filter, and appear in weakly monotonic time order.

Reply:
@lihui815 Hold on. Seek is not a filter. 
At least it's supposed to be a one time command for adjusting pointer to current message with requested time.
Am I missed something?

Reply:
If user calling `has_next()` it should return if bag really has next message independently from `seek()` operation.
Per your explanation logic will  get broken and will have side effect not as we expected.

Reply:
@lihui815 @emersonknapp IMO #842 is a completely separate issue and is matter of discussion how this possibility of writing out of ordered messages could be workarounded or even fixed. But it certainly shouldn't be addressed in this PR using `seek()` API as filter.

Playback must replay in exact order as messages was recorded. **We need this condition for deterministic replay.** Otherwise it will be difficult to reproduce exact behavior of the entire system if player will do some internal reordering of the messages which he was not asked for.
If messages were recorder out of order this is problem of recorder or post processing tool.

Reply:
This conversation is off topic for this feature - if you have a specific issue with the current behavior (order by timestamp) - open a new issue to discuss.

Reply:
@emersonknapp @lihui815 Of course I can create follow up PR. Just want to let you know that current implementation will introduce **bug**.
With current implementation in case when we would have second DB file with first message in it with timestamp early then the in last message in the first file (as described in #842) we are going to skip this first message in second DB file. 
Not playing message just because it was recorder out of order will be considered as a bug and will lead to the non-deterministic replay. 
Just imagine if it would be some key frame for some algorithm or triggering message and we will not play it.

IMO seek `seek_time_` should be updated only from `seek(timestamp)` API.

Reply:
@lihui815 Sorry, but I am still not get it.
Why introduce bug in first place and link it to some issue for further consideration, rather then just fix it on code review?
It will be obviously a buggy behavior with skipping message. 
I would disagree with this
>skipping messages that appear out of order in next file - yes i considered this, however this is something i would wrap into issue of #842 because it's part of the generic issue of messages being able to be read out of time order

Before this PR behavior was consistent and we didn't skip messages which was recorded out of order. 

I agree that seek operation doesn't make sense if messages stored in bag in out of timestamp order.
And 
> so I'd rather not make implementation workarounds for this

This is exactly what I am asking for. Please don't try to workaround situation with messages stored out of order in this PR.

Reply:
Currently we don't have real cases with situation when messages stored out of timestamp order in different files. 
Issue #842 describe potential problem if someone will potentially try to incorrectly use direct API when saving messages to bag files.

Reply:
Agree changing behavior not best practice. removed `seek_time_` updating in `SequentialReader->read_next()`, and clarified that `SequentialReader->seek(t)` will just ensure that all subsequent reads will return all messages at or after time `t`, and will not guarantee time order.